src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.cs

```diff
@@ -58,10 +58,11 @@ private static FileStreamStrategy WrapIfDerivedType(FileStream fileStream, FileS
             e is NotSupportedException ||
             (e is ArgumentException && !(e is ArgumentNullException));
 
-        internal static bool ShouldPreallocate(long preallocationSize, FileAccess access, FileMode mode)
+        internal static bool ShouldPreallocate(long preallocationSize, FileAccess access, FileMode mode, SafeFileHandle fileHandle)
             => preallocationSize > 0
                && (access & FileAccess.Write) != 0
-               && mode != FileMode.Open && mode != FileMode.Append;
+               && mode != FileMode.Open && mode != FileMode.Append
+               && (mode != FileMode.OpenOrCreate || fileHandle.CanSeek && RandomAccess.GetFileLength(fileHandle) == 0); // allow to extend only new files
```

Comment:
Nit: please add parens around the latter expression... precedence between || and && is hard to remember.

Reply:
Just because a file length is 0 doesn't mean it's new.